mypyc/irbuild/specialize.py

```diff
@@ -356,3 +358,53 @@ def translate_dict_setdefault(
                               [callee_dict, key_val, data_type],
                               expr.line)
     return None
+
+
+@specialize_function('format', str_rprimitive)
+def translate_str_format(
+        builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
+    if (isinstance(callee, MemberExpr) and isinstance(callee.expr, StrExpr)
+            and expr.arg_kinds.count(ARG_POS) == len(expr.arg_kinds)):
+
+        format_str = callee.expr.value
+        if not can_optimize_format(format_str):
+            return None
+
+        format_str = format_str.replace("{{", "{")
+        format_str = format_str.replace("}}", "}")
+
+        literals = format_str.split("{}")
+        variables = [builder.accept(x) if is_str_rprimitive(builder.node_type(x))
+                     else builder.call_c(str_op, [builder.accept(x)], expr.line)
+                     for x in expr.args]
+
+        # The first parameter is the total size of the following PyObject* merged from
+        # two lists alternatively.
+        result_list: List[Value] = [Integer(0, c_int_rprimitive)]
+        for a, b in zip(literals, variables):
+            if a:
+                result_list.append(builder.load_str(a))
+            result_list.append(b)
+        # The str.split() always generates one more element
+        if literals[-1]:
+            result_list.append(builder.load_str(literals[-1]))
+
+        result_list[0] = Integer(len(result_list) - 1, c_int_rprimitive)
+        return builder.call_c(str_build_op, result_list, expr.line)
+    return None
+
+
+def can_optimize_format(format_str: str) -> bool:
```

Comment:
Please add unit tests for this, since it has pretty tricky logic. Also consider various error cases and edge cases.